client/asset/eth/eth.go

```diff
@@ -3886,15 +3889,266 @@ func checkTxStatus(receipt *types.Receipt, gasLimit uint64) error {
+
+	if err := createWallet(&asset.CreateWalletParams{
+		Type:     walletTypeRPC,
+		Seed:     seed,
+		Pass:     pw,
+		Settings: map[string]string{providersKey: provider},
+		DataDir:  walletDir,
+		Net:      net,
+		Logger:   log,
+	}, true); err != nil {
+		return nil, nil, fmt.Errorf("error creating initiator wallet: %v", err)
+	}
+
+	ethCfg, err := ethChainConfig(net)
+	if err != nil {
+		return nil, nil, err
+	}
+	chainConfig := ethCfg.Genesis.Config
+
+	cl, err := newMultiRPCClient(walletDir, []string{provider}, log, chainConfig, big.NewInt(chainIDs[net]), net)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error opening initiator rpc client: %v", err)
+	}
+
+	if err = cl.connect(ctx); err != nil {
+		return nil, nil, fmt.Errorf("error connecting: %v", err)
+	}
+	defer cl.shutdown()
+
+	if err = cl.unlock(string(pw)); err != nil {
+		return nil, nil, fmt.Errorf("error unlocking initiator node: %v", err)
+	}
+
+	var c contractor
+	if assetID == BipID {
+		ctor := contractorConstructors[contractVer]
+		if ctor == nil {
+			return nil, nil, fmt.Errorf("no contractor constructor for eth contract version %d", contractVer)
+		}
+		c, err = ctor(net, cl.address(), cl.contractBackend())
+		if err != nil {
+			return nil, nil, fmt.Errorf("contractor constructor error: %v", err)
+		}
+	} else {
+		ctor := tokenContractorConstructors[contractVer]
+		if ctor == nil {
+			return nil, nil, fmt.Errorf("no token contractor constructor for eth contract version %d", contractVer)
+		}
+		c, err = ctor(net, assetID, cl.address(), cl.contractBackend())
+		if err != nil {
+			return nil, nil, fmt.Errorf("token contractor constructor error: %v", err)
+		}
+	}
+	return cl, c, nil
+}
+
+// waitForConfirmation waits for the specified transaction to have > 0
+// confirmations.
+func waitForConfirmation(ctx context.Context, cl ethFetcher, txHash common.Hash) error {
+	bestHdr, err := cl.bestHeader(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting best header: %w", err)
+	}
+	ticker := time.NewTicker(blockTicker)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			hdr, _ := cl.bestHeader(ctx)
+			if hdr != nil && hdr.Number.Cmp(bestHdr.Number) > 0 {
+				bestHdr = hdr
+				confs, err := cl.transactionConfirmations(ctx, txHash)
+				if err != nil {
+					if !errors.Is(err, asset.CoinNotFoundError) {
+						return fmt.Errorf("error getting transaction confirmations: %v", err)
+					}
+					continue
+				}
+				if confs > 0 {
+					return nil
+				}
+			}
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// runSimnetMiner starts a gouroutine to generate a simnet block every 5 seconds
+// until the ctx is canceled.
+func runSimnetMiner(ctx context.Context, log dex.Logger) {
+	log.Infof("Starting the simnet miner")
+	go func() {
+		tick := time.NewTicker(time.Second * 5)
+		u, _ := user.Current()
+		for {
+			select {
+			case <-tick.C:
+				log.Debugf("Mining a simnet block")
+				mine := exec.CommandContext(ctx, "./mine-alpha", "1")
```

Comment:
Oh really? I thought that was just kinda random because you can only specify to start/stop the miner with eth. But it is (almost?) always 2 blocks. Is that a geth thing?

Reply:
iirc we set the block time to one second.

Geth has no way to say "mine one block."

So we mining.Start, then mining.Stop later. If we don't wait one second before we stop, it might not mine one at all. If we wait a second before calling mining.Stop, it might mine another or it might not. So, stopping after it at least mined one was the safest bet, and not shooting for one.

Sorry, not really a precision of two, just the first block can't guarantee one, and the rest are also sort of one off at times.